proposals/nnnn-inferring-senable-for-methods.md

```diff
@@ -0,0 +1,221 @@
+
+
+```
+struct S: P {
+  func f() { ... }
+}
+
+g(S.f) // Converting non-sendable function value to '@Sendable (S) -> (() -> Void)' may introduce data races
+```
+
+
+We can work around this by wrapping our unapplied function in a Sendable closure.  
+
+```
+// S.f($0) == S.f()
+g({ @Sendable **in** S.f($0) })
+```
+
+
+This is a lot of churn to get the expected behavior. The compiler should preserve `@Sendable` in the type signature instead.
+
+## Proposed solution
+
+We propose the compiler should automatically employ `@Sendable`  to functions that cannot capture non-Sendable states. This includes partially-applied and unapplied instance methods of `Sendable` types, as well as non-local functions. Additionally, it should be disallowed to utilize `@Sendable` on instance methods of non-`Sendable` types.
+
+For a function, the `@Sendable` attribute primarily influences the kinds of values that can be captured by the function. But methods of a nominal type do not capture anything but the object instance itself. Semantically, a method can be thought of as being represented by the following functions:
+
+
+```
+// Pseudo-code declaration of a Nominal Type:
+type NominalType {
+  func method(ArgType) -> ReturnType { /* body of method */ }
+}
+
+// Can desugar to these two global functions:
+func NominalType_method_partiallyAppliedTo(_ obj: NominalType) -> ((ArgType) -> ReturnType) {
+  let inner = { [obj] (_ arg1: ArgType) -> ReturnType in
+    return NominalType_method(obj, arg1)
+  }
+  return inner
+}
+// The actual method call
+func NominalType_method(_ self: NominalType, _ arg1: ArgType) -> ReturnType {
+  /* body of method */
+}
+```
+
+Thus, the only way a partially-applied method can be `@Sendable` is if the `inner` closure were `@Sendable`, which is true if and only if the nominal type conforms to `Sendable`.
+
+
+```
+type NominalType : Sendable {
+  func method(ArgType) -> ReturnType { /* body of method */ }
+}
+```
+
+For example, by declaring the following type `Sendable`, the partial and unapplied function values of the type would have implied Sendability and the following code would compile with no errors.
+```
+struct User : Sendable {
+  func updatePassword (new: String, old:String) -> Bool {
+    /* update password*/ 
+    return true
+  }
+}
+
+let unapplied: @Sendable (User) -> ((String, String) → Bool) = User.updatePassword // no error
+
+let partial: @Sendable (String, String) -> Bool = User().updatePassword // no error
+```
+
+
+
+## Detailed design
+
+This proposal includes four changes to `Sendable` behavior.
+
+The first two are what we just discussed regarding partial and unapplied methods.
+
+```
+struct User : Sendable {
+  var address
+  var password
+  
+  func changeAddress () {/*do work*/ }
+}
+```
+
+1. The inference of `@Sendable` for unapplied references to methods of a Sendable type. 
+
+```
+let unapplied : @Sendable (User) → ((String, String) → Void) = User.changeAddress // no error
```

Comment:
First of all, there are some stray arrows to fix:
```suggestion
let unapplied : @Sendable (User) -> ((String, String) -> Void) = User.changeAddress // no error
```

However, I'm not sure if you've actually gotten the right rule here. It seems to me that (a) the unapplied method itself should always be `@Sendable`—whether or not the type is `Sendable`—because it doesn't capture any state, and (b) the partially applied method it returns should be `@Sendable` if the type is `Sendable`. That rule would look like this:
```suggestion
let unapplied : @Sendable (User) -> (@Sendable (String, String) -> Void) = User.changeAddress // no error
```

If you do go with this, it might make sense to reverse changes 1 and 2 so you can refer back to the partially-applied rule when talking about the result type of the unapplied method.

(As an aside, it might be helpful to also include a non-Sendable type in your examples and show how the new rules apply to it, to demonstrate how they're different.)

Reply:
Thinking about this.

Reply:
I see what you're saying.  Yes that is more correct. I've updated the implementation to reflect this. Will update the proposal as well.